source/common/ssl/context_impl.cc

```diff
@@ -370,6 +383,133 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
+  RELEASE_ASSERT(rc == 1);
+
+  // Include SERVER_SESSION_ID_CONTEXT so that if all the other verify-settings are unset
+  // we have a deterministic value.
+  rc = EVP_DigestUpdate(&md, &SERVER_SESSION_ID_CONTEXT, sizeof(SERVER_SESSION_ID_CONTEXT));
+  RELEASE_ASSERT(rc == 1);
+
+  if (ca_cert_ != nullptr) {
+    rc = X509_digest(ca_cert_.get(), EVP_sha256(), session_context_buf, &session_context_len);
+    RELEASE_ASSERT(rc == 1 && session_context_len == SHA256_DIGEST_LENGTH);
+    rc = EVP_DigestUpdate(&md, session_context_buf, session_context_len);
+    RELEASE_ASSERT(rc == 1);
+  }
+
+  for (const std::string& name : verify_subject_alt_name_list_) {
+    rc = EVP_DigestUpdate(&md, name.data(), name.size());
+    RELEASE_ASSERT(rc == 1);
+  }
+
+  rc = EVP_DigestUpdate(&md, verify_certificate_hash_.data(),
+                        verify_certificate_hash_.size() *
+                            sizeof(decltype(verify_certificate_hash_)::value_type));
+  RELEASE_ASSERT(rc == 1);
+
+  rc = EVP_DigestFinal(&md, session_context_buf, &session_context_len);
+  RELEASE_ASSERT(rc == 1);
+  rc = SSL_CTX_set_session_id_context(ctx_.get(), session_context_buf, session_context_len);
+  RELEASE_ASSERT(rc == 1);
+}
+
+int ServerContextImpl::sessionTicketProcess(SSL*, uint8_t* key_name, uint8_t* iv,
+                                            EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx, int encrypt) {
+  const EVP_MD* hmac = EVP_sha256();
+  const EVP_CIPHER* cipher = EVP_aes_256_cbc();
+
+  if (encrypt == 1) {
+    // Encrypt
+    RELEASE_ASSERT(session_ticket_keys_.size() >= 1);
+    const SessionTicketKey& key = session_ticket_keys_.front();
+
+    static_assert(std::tuple_size<decltype(key.name)>::value == SSL_TICKET_KEY_NAME_LEN,
+                  "Expected key.name length");
+    std::copy_n(key.name.begin(), SSL_TICKET_KEY_NAME_LEN, key_name);
+
+    int got_rand_bytes = RAND_bytes(iv, EVP_CIPHER_iv_length(cipher));
+    RELEASE_ASSERT(got_rand_bytes);
+
+    RELEASE_ASSERT(key.aes_key.size() == EVP_CIPHER_key_length(cipher));
+    if (!EVP_EncryptInit_ex(ctx, cipher, nullptr, key.aes_key.data(), iv)) {
+      return -1;
+    }
+
+    if (!HMAC_Init_ex(hmac_ctx, key.hmac_key.data(), key.hmac_key.size(), hmac, nullptr)) {
+      return -1;
+    }
+
+    return 1; // success
+  } else {
+    // Decrypt
+    bool is_enc_key = true; // first element is the encryption key
+    for (const SessionTicketKey& key : session_ticket_keys_) {
+      static_assert(std::tuple_size<decltype(key.name)>::value == SSL_TICKET_KEY_NAME_LEN,
+                    "Expected key.name length");
+      if (std::equal(key.name.begin(), key.name.end(), key_name)) {
+        if (!HMAC_Init_ex(hmac_ctx, key.hmac_key.data(), key.hmac_key.size(), hmac, nullptr)) {
+          return 0;
+        }
+
+        RELEASE_ASSERT(key.aes_key.size() == EVP_CIPHER_key_length(cipher));
+        if (!EVP_DecryptInit_ex(ctx, cipher, nullptr, key.aes_key.data(), iv)) {
+          return 0;
```

Comment:
This should be `return -1;`.

Reply:
Same for this.  Can it happen in any cases we care about?  If not, I can just ASSERT success.